rust/private/rustc.bzl

```diff
@@ -822,6 +830,20 @@ def _add_native_link_flags(args, dep_info, crate_type, cc_toolchain, feature_con
     args.add_all(dep_info.transitive_dylibs, map_each = get_lib_name, format_each = "-ldylib=%s")
     args.add_all(dep_info.transitive_staticlibs, map_each = get_lib_name, format_each = "-lstatic=%s")
 
+    # alwayslink=True libraries need special handling so the alwayslink args
+    # make it to the linker. Eventually we might be able to use a first-class
+    # rustc flag for this (https://github.com/rust-lang/cargo/issues/7586),
+    # but until then we have to do it this way.
+    for lib in dep_info.transitive_staticlibs_alwayslink.to_list():
```

Comment:
I know .to_list() is called in way too many places in this repo, but that's something we will have to fix. Could you instead extract a function that produces the right flags and use args.add_all(libs, map_each=...)? That would defer the depset flattening till the action execution phase thus speed up analysis time and decrease memory use. Thank you!

Reply:
As far as I can starlark, this can't be done: the map_each function needs information about the OS in play or it can't specify the right format linker flag, and there's no way to set up a partial in starlark. It also looks like there's nothing like __call__ from Python here, so I've tried to structure things as similarly to what you've asked for here so that if I'm wrong it's an easy follow-up but for now this isn't technically done.

Reply:
You could have 3 different functions and choose one of them depending on the OS when constructing args, right? Or am I missing something?

Reply:
Took a wee bit more refactoring than that, but it's fixed.